prov/util/src/util_smrn.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2017 Cray Inc. All rights reserved.
+ * Copyright (c) 2017 Intel Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <ofi_mr.h>
+#include <inttypes.h>
+
+#define UTIL_SMRN_GET_PROV(smrn)	((smrn)->attr.prov)
+
+static struct util_smrn global_smrn;
+
+static inline int
+util_check_smrn_attr_sanity(struct util_smrn_attr *attr)
+{
+	/* callbacks and provider must be provided */
+	if (!attr || !attr->prov ||
+	    !attr->init || !attr->open || !attr->close ||
+	    !attr->monitor || !attr->unmonitor || !attr->get_event)
+		return -FI_EINVAL;
+
+	/* valid otherwise */
+	return FI_SUCCESS;
+}
+
+int ofi_util_smrn_init(struct util_smrn_attr *attr)
+{
+	int ret;
+
+	ret = util_check_smrn_attr_sanity(attr);
+	if (ret)
+		return ret;
+
+	memcpy(&global_smrn.attr, attr, sizeof(*attr));
+	fastlock_init(&global_smrn.lock);
+	global_smrn.references = 0;
+	dlist_init(&global_smrn.rq_head);
+
+	return global_smrn.attr.init();
```

Comment:
This was done in mind with the idea of abstracting it later. Dmitry has already done some of the abstraction, but it can be further abstracted. This function call should be redone.